src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -24,10 +24,96 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 #include "gcinfo.h"
 #include "gcinfoencoder.h"
 
+//------------------------------------------------------------------------
+// genIsTableDrivenHWIntrinsic:
+//
+// Arguments:
+//    category - category of a HW intrinsic
+//
+// Return Value:
+//    returns true if this category can be table-driven in CodeGen
+//
+bool CodeGen::genIsTableDrivenHWIntrinsic(HWIntrinsicCategory category)
+{
+    // TODO - make more categories to the table-driven framework
+    HWIntrinsicCategory drivableCategory    = HW_Category_SimpleSIMD;
+    HWIntrinsicCategory nondrivableCategory = HW_Category_Special;
+    return (category & drivableCategory) != 0 && (category & HW_Category_Special) == 0;
+}
+
 void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
 {
-    NamedIntrinsic intrinsicID = node->gtHWIntrinsicId;
-    InstructionSet isa         = compiler->isaOfHWIntrinsic(intrinsicID);
+    NamedIntrinsic      intrinsicID = node->gtHWIntrinsicId;
+    InstructionSet      isa         = compiler->isaOfHWIntrinsic(intrinsicID);
+    HWIntrinsicCategory category    = compiler->categoryOfHWIntrinsic(intrinsicID);
+
+    assert((category & HW_Category_NoCodeGen) == 0);
+
+    if (genIsTableDrivenHWIntrinsic(category))
+    {
+
+        GenTree*  op1        = node->gtGetOp1();
+        GenTree*  op2        = node->gtGetOp2();
+        regNumber targetReg  = node->gtRegNum;
+        var_types targetType = node->TypeGet();
+        var_types baseType   = node->gtSIMDBaseType;
+        int       numArgs    = node->gtNumArgs;
+        regNumber op1Reg     = REG_NA;
+        regNumber op2Reg     = REG_NA;
+        emitter*  emit       = getEmitter();
+
+        assert(numArgs >= 0);
+        instruction ins = compiler->insOfHWIntrinsic(intrinsicID, baseType);
+        assert(ins != INS_invalid);
+        unsigned simdSize = node->gtSIMDSize;
+        assert(simdSize != 0);
+
+        switch (numArgs)
+        {
+            case 1:
+                genConsumeOperands(node);
+                op1Reg = op1->gtRegNum;
+                emit->emitIns_R_R(ins, (emitAttr)simdSize, targetReg, op1Reg);
+                break;
+            case 2:
+                genConsumeOperands(node);
+                op1Reg = op1->gtRegNum;
+                op2Reg = op2->gtRegNum;
+                emit->emitIns_SIMD_R_R_R(ins, targetReg, op1Reg, op2Reg, (emitAttr)simdSize);
+                break;
+            case 3:
+            {
+                assert(op1->OperIsList() && op1->gtGetOp2()->OperIsList() && op1->gtGetOp2()->gtGetOp2()->OperIsList());
```

Comment:
nit: I tend to prefer `&&` asserts to be split into multiple asserts (one per line). It makes it significantly easier to debug as you can determine which assert in the chain actually failed.